### PR TITLE
[Merged by Bors] - fix diagnostic length for asset count

### LINKED ISSUE #2156: Multiple issues when using AssetCountDiagnosticPlugin
## Bevy version

0.5.0

## Operating system & version

Manjaro 21.04

## What you did

Attempt to use the AssetCountDiagnosticsPlugin with a Texture or Mesh.

## What you expected to happen

I expected this to work without error
```rust
use bevy::prelude::*;
use bevy::asset::diagnostic::AssetCountDiagnosticsPlugin;

fn main() {
    App::build()
        .add_plugins(DefaultPlugins)
        .add_plugin(AssetCountDiagnosticsPlugin::<Texture>::default())
        .run()
}
```
## What actually happened

I receive this error in the console
```
May 13 11:31:16.496  WARN bevy_diagnostic::diagnostic: Diagnostic "asset_count bevy_render::texture::texture::Texture" has name longer than 32 characters, and so might overflow in the LogDiagnosticsPlugin Consider using a shorter name.
```

I may be wrong but since the 'name' it is complaining about is the full path of the object, can this 'name' be changed from outside bevy?

## Additional Error

Another thing I've noticed is I'm unable to to create a AssetCountDiagnosticsPlugin for Mesh assets.

Since Mesh doesn't implement Default it fails to compile

```
   Compiling asset-count v0.1.0 (/home/patrick/Projects/asset-count)
error[E0599]: the function or associated item `default` exists for struct `AssetCountDiagnosticsPlugin<bevy::prelude::Mesh>`, but its trait bounds were not satisfied
   --> src/main.rs:7:58
    |
7   |         .add_plugin(AssetCountDiagnosticsPlugin::<Mesh>::default())
    |                                                          ^^^^^^^ function or associated item cannot be called on `AssetCountDiagnosticsPlugin<bevy::prelude::Mesh>` due to unsatisfied trait bounds
    |
   ::: /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_asset-0.5.0/src/diagnostic/asset_count_diagnostics_plugin.rs:8:1
    |
8   | pub struct AssetCountDiagnosticsPlugin<T: Asset> {
    | ------------------------------------------------ doesn't satisfy `_: Default`
    |
   ::: /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_render-0.5.0/src/mesh/mesh.rs:209:1
    |
209 | pub struct Mesh {
    | --------------- doesn't satisfy `bevy::prelude::Mesh: Default`
    |
    = note: the following trait bounds were not satisfied:
            `bevy::prelude::Mesh: Default`
            which is required by `AssetCountDiagnosticsPlugin<bevy::prelude::Mesh>: Default`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
error: could not compile `asset-count`

To learn more, run the command again with --verbose.
```

By looking at the source of AssetCountDiagnosticsPlugin it appears that the default method is only implemented where T: Default. But it should be implemented for any T I believe

### PR DESCRIPTION
**Pull request by mockersf**
fixes #2156 
limit the diagnostic name to `MAX_DIAGNOSTIC_NAME_WIDTH` length

### CODE REVIEW

**Comment by NathanSWard** [comment]
Please correct me if I'm wrong, but the length of the string is still capped at `32`, correct? So this won't actually change any of the behavior from the first problem #2156 mentioned.

File: crates/bevy_diagnostic/src/lib.rs [hunk 1]
```diff
@@ -18,3 +18,7 @@ impl Plugin for DiagnosticsPlugin {
         app.init_resource::<Diagnostics>();
     }
 }
+
+/// The width which diagnostic names will be printed as
+/// Plugin names should not be longer than this value
+pub const MAX_DIAGNOSTIC_NAME_WIDTH: usize = 32;
```
Comment (NathanSWard) [inline crates/bevy_diagnostic/src/lib.rs:24]: What's the purpose of having this limit in the first place?

Comment (mockersf) [reply crates/bevy_diagnostic/src/lib.rs:24]: it came in #1505

Comment (mockersf) [reply crates/bevy_diagnostic/src/lib.rs:24]: it was to clean up how it's displayed, and it's an issue only for this diagnostics as all other have a fixed name

File: crates/bevy_asset/src/diagnostic/asset_count_diagnostics_plugin.rs [hunk 2]
```diff
@@ -29,9 +29,20 @@ impl<T: Asset> AssetCountDiagnosticsPlugin<T> {
...
+        let max_length = MAX_DIAGNOSTIC_NAME_WIDTH - "asset_count ".len();
         diagnostics.add(Diagnostic::new(
             Self::diagnostic_id(),
-            format!("asset_count {}", std::any::type_name::<T>()),
+            format!(
+                "asset_count {}",
+                if asset_type_name.len() > max_length {
+                    asset_type_name
+                        .split_at(asset_type_name.len() - max_length + 1)
+                        .1
```
Comment (NathanSWard) [inline crates/bevy_asset/src/diagnostic/asset_count_diagnostics_plugin.rs:41]:
```suggestion
                        .0
```
This should be the first `&str` returned by `split_at` (assuming you want the first part of the type name).

Comment (mockersf) [reply crates/bevy_asset/src/diagnostic/asset_count_diagnostics_plugin.rs:41]: I want the second, the first is most likely to be part of the path: `asset_count bevy_render::texture::texture::Texture`

**Review by NathanSWard — APPROVED** [review]

**Comment by cart** [comment]
Looks good to me!

**Pull request closed without merging**